apis/v1/gateway_types.go

```diff
@@ -188,6 +188,7 @@ type GatewaySpec struct {
 	// +kubebuilder:validation:MaxItems=64
 	// +kubebuilder:validation:XValidation:message="tls must be specified for protocols ['HTTPS', 'TLS']",rule="self.all(l, l.protocol in ['HTTPS', 'TLS'] ? has(l.tls) : true)"
 	// +kubebuilder:validation:XValidation:message="tls must not be specified for protocols ['HTTP', 'TCP', 'UDP']",rule="self.all(l, l.protocol in ['HTTP', 'TCP', 'UDP'] ? !has(l.tls) : true)"
+	// +kubebuilder:validation:XValidation:message="tls mode must be Terminate for protocol HTTPS",rule="self.all(l, l.protocol==HTTPS ? l.tls==Terminate)"
```

Comment:
This is surprisingly tricky. I agree that if Listener protocol is "HTTPS", we've said that the only valid TLS Mode is "Terminate" ([docs](https://gateway-api.sigs.k8s.io/guides/tls/#clientserver-and-tls)). Of course that is also the [default value for TLS mode](https://github.com/kubernetes-sigs/gateway-api/blob/a6c0e9ce5dd38cc3a065a672d4132f70a63cf087/apis/v1/gateway_types.go#L394), and I'm not sure whether CEL or defaulting comes first, I think it's defaulting, but we need to have some tests that cover that case.

For example, I think the following example should still be valid (relying on defaulting) if we remove the mode from the config: https://github.com/kubernetes-sigs/gateway-api/blob/a6c0e9ce5dd38cc3a065a672d4132f70a63cf087/examples/standard/simple-http-https/gateway.yaml#L1-L20

It would be good to add two variations of that example - one alongside this just without mode specified (this would be valid) and a separate copy in `hack/invalid-examples` that sets mode to "Passthrough" (invalid).

As far as the overall CEL, I think you add least need to check for TLS Mode specifically, but this may need further adjustments depending on defaulting behavior.

```suggestion
	// +kubebuilder:validation:XValidation:message="tls mode must be Terminate for protocol HTTPS",rule="self.all(l, l.protocol==HTTPS ? l.tls.Mode==Terminate)"
```

It's also challenging to make a backwards incompatible change to validation on a GA API like this, but I believe this is the correct decision. In this case I believe it falls under "Correcting validation" in our [versioning guidelines](https://gateway-api.sigs.k8s.io/concepts/versioning/#patch-version-eg-v040-v041). Assuming we can get this right, this will only invalidate HTTPS listeners with TLS Mode set to "Passthrough" which is a state the spec has long considered invalid.

Reply:
Thanks Rob for the clarification regarding supportability!

I wonder if we need the "else" clause of CEL ternary expression...hopefully the test run should reveal this.

Reply:
Oh yep, good call, would also need an `:` in the ternary.